components/FilterableNumbersList.js

```diff
@@ -1,18 +1,47 @@
 import * as React from 'react';
 import { NumbersList } from './NumbersList';
+import { withRouter } from 'next/router';
+import { DebounceInput } from 'react-debounce-input';
+
+class _FilterableNumbersList extends React.Component {
+  state = {
+    query: this.props.router.query.q
+  };
+
+  updateRouter = query => {
+    this.props.router.replace({
+      pathname: '/',
+      query: { q: query }
+    });
+  };
+
+  updateQuery = e => {
+    const value = e.target.value;
+    this.setState({ query: value }, () => {
+      this.updateRouter(value);
```

Comment:
The goal of splitting up the state of the search box and the routing update is to keep the state of the input from being slowed down by routing, which will be updated asynchronously